### ISSUE TITLE
Run the ".run_all.sh" file, get stuck, and don't execute any further.

### ISSUE THREAD

**Issue by cxk1686** [nkallima/sim-universal-construction#53]
Hello, I have a question I would like to ask you: that is, when I run ".run_all.sh", I encountered a stuck phenomenon. Excuse me, how should this be resolved? thanks
I am using an ARM64 system with a raspberry pi.
```
pi@raspberrypi:/opt/sim-universal-construction $ cat /proc/cpuinfo
processor	: 0
BogoMIPS	: 108.00
Features	: fp asimd evtstrm crc32 cpuid
CPU implementer	: 0x41
CPU architecture: 8
CPU variant	: 0x0
CPU part	: 0xd08
CPU revision	: 3

processor	: 1
BogoMIPS	: 108.00
Features	: fp asimd evtstrm crc32 cpuid
CPU implementer	: 0x41
CPU architecture: 8
CPU variant	: 0x0
CPU part	: 0xd08
CPU revision	: 3

processor	: 2
BogoMIPS	: 108.00
Features	: fp asimd evtstrm crc32 cpuid
CPU implementer	: 0x41
CPU architecture: 8
CPU variant	: 0x0
CPU part	: 0xd08
CPU revision	: 3

processor	: 3
BogoMIPS	: 108.00
Features	: fp asimd evtstrm crc32 cpuid
CPU implementer	: 0x41
CPU architecture: 8
CPU variant	: 0x0
CPU part	: 0xd08
CPU revision	: 3

Hardware	: BCM2835
Revision	: d03141
Serial		: 10000000ad931999
Model		: Raspberry Pi Compute Module 4 Rev 1.1
```
```
root@raspberrypi:/opt/sim-universal-construction# . run_all.sh 
Executing activesetbench.run (1/32) ...Done
Executing ccqueuebench.run (2/32) ...Done
Executing ccstackbench.run (3/32) ...Done
Executing ccsynchbench.run (4/32)
```

**Comment by nkallima** [comment]
@cxk1686 Thanks for reporting this.
Could you give more details about the Linux distro that you are using? Is a 64-bit distro or a 32-bit?
Moreover, please check if `validate.sh` runs successfully or not.

**Comment by cxk1686** [comment]
> @cxk1686 Thanks for reporting this. Could you give more details about the Linux distro that you are using? Is a 64-bit distro or a 32-bit? Moreover, please check if `validate.sh` runs successfully or not.

hello, When I run the validate.sh file, I also get stuck.
`root@raspberrypi:/opt/sim-universal-construction# . validate.sh 
Compiling the sources...						[ PASS ]

Validating for   1 thread(s)
============================
Validating ccsynchbench.run     					[ PASS ]
Validating dsmsynchbench.run    					[ PASS ]
Validating hsynchbench.run      					[ PASS ]
Validating oscibench.run        					[ PASS ]
Validating simbench.run         					[ PASS ]
Validating fcbench.run          					[ PASS ]
Validating oyamabench.run       					[ PASS ]
Validating mcsbench.run         					[ PASS ]
Validating clhbench.run         					[ PASS ]
Validating pthreadsbench.run    					[ PASS ]
Validating fadbench.run         					[ PASS ]
Validating ccstackbench.run     					[ PASS ]
Validating clhstackbench.run    					[ PASS ]
Validating dsmstackbench.run    					[ PASS ]
Validating hstackbench.run      					[ PASS ]
Validating oscistackbench.run   					[ PASS ]
Validating simstackbench.run    					[ PASS ]
Validating fcstackbench.run     					[ PASS ]
Validating ccqueuebench.run     					[ PASS ]
Validating clhqueuebench.run    					[ PASS ]
Validating dsmqueuebench.run    					[ PASS ]
Validating hqueuebench.run      					[ PASS ]
Validating osciqueuebench.run   					[ PASS ]
Validating simqueuebench.run    					[ PASS ]
Validating fcqueuebench.run     					[ PASS ]
Validating lcrqbench.run        					[ PASS ]
Validating clhhashbench.run     					[ PASS ]
Validating dsmhashbench.run     					[ PASS ]

Validating for   2 thread(s)
============================
Validating ccsynchbench.run`

**Comment by nkallima** [comment]
@cxk1686, thanks for the additional info.

Unfortunately, I currently do not have access to a raspberry PI board. Thus, I'm not able to reproduce this.
However, it seems that `ccsynchbench.run` does not finish its execution for 2 or more threads.
Could you please check if there is any other benchmark that behaves similarly, e.g. `dsmsynchbench.run`, `fadbench.run`, etc?
You could execute any benchmark by running `./build/bin/benchmark_binary.run -t 2', where benchmark_binary.run is the binary of the benchmark and `-t` is the argument for controlling the number of threads.

Cheers

**Comment by cxk1686** [comment]
> @cxk1686, thanks for the additional info.
> 
> Unfortunately, I currently do not have access to a raspberry PI board. Thus, I'm not able to reproduce this. However, it seems that `ccsynchbench.run` does not finish its execution for 2 or more threads. Could you please check if there is any other benchmark that behaves similarly, e.g. `dsmsynchbench.run`, `fadbench.run`, etc? You could execute any benchmark by running `./build/bin/benchmark_binary.run -t 2', where benchmark_binary.run is the binary of the benchmark and `-t` is the argument for controlling the number of threads.
> 
> Cheers

@edatsika OK, I will try it, but I changed the RUNS_PER_THREAD=100000 in the validate.h file to a smaller value, such as RUNS_PER_THREAD=800, and the file runs successfully.

Thanks

**Comment by cxk1686** [comment]
I changed the value of SYNCH_RUNS and made it smaller, and run_all.sh ran successfully. 
Excuse me, does this mean that my test environment configuration is low?

**Comment by nkallima** [comment]
@cxk1686,

Try to increase RUNS_PER_THREAD in relatively steps, (for example, doubling its value in every step).
In case that your system is slow, the validation test would need more time to finish in each test.
In case that there is some race, then the validation test may fail to finish its execution even in short tests.

**Comment by cxk1686** [comment]
> @cxk1686,
> 
> Try to increase RUNS_PER_THREAD in small relatively steps, (for example, doubling its value in every step). In case that your system is slow, the validation test would need more time to finish in each test. In case that there is some race, then the validation test may fail to finish its execution even in short tests.

ok, i'll try what you said, thanks a lot, cheers.

**Comment by cxk1686** [comment]
> @cxk1686,
> 
> Try to increase RUNS_PER_THREAD in small relatively steps, (for example, doubling its value in every step). In case that your system is slow, the validation test would need more time to finish in each test. In case that there is some race, then the validation test may fail to finish its execution even in short tests.

ok, i'll try what you said, thanks a lot, cheers.

**Comment by nkallima** [comment]
Hello k1686,

Today, I got access to a Raspbian aarch64 on a Pi3.
It seems that `validate.sh` and `run_all.sh` finish their execution without any errors.

From your `/proc/cpuinfo` report, I see that you are running on a BCM2835, which is a 32-bit processor?

**Comment by cxk1686** [comment]
@nkallima 
My raspberry pi has a 64-bit CPU.
```
Processors
SOC/Package Information	
ARM Cortex-A72 r0p3 (AArch64)	1500.00 MHz
ARM Cortex-A72 r0p3 (AArch64)	1500.00 MHz
ARM Cortex-A72 r0p3 (AArch64)	1500.00 MHz
ARM Cortex-A72 r0p3 (AArch64)	1500.00 MHz
```

**Comment by cxk1686** [comment]
Hello @nkallima 
I want to add C++ code to the project, but I am having difficulty changing the Makefile and Makefile.generic, can you help me?

```
x86_64:
	make -f Makefile.generic -j $(NPROC) COMPILER=$(CXX) LDLIBS='$(LIBS)' CXXFLAGS='-Wall -std=c++11 -Ofast -fPIC -flto -ftree-vectorize -finline-functions -march=native -mtune=native'
unknown:
	make -f Makefile.generic -j $(NPROC) COMPILER=$(CXX) LDLIBS='$(LIBS)'  CFLAGS='-std=c++11 -Ofast -fPIC -flto'
```

```
COMPILER       = g++
AR             = gcc-ar
LDLIBS         = -lpthread -lm -lnuma
CFLAGS         =
ARGS           =
D_ARGS         =
INSTALL_INFO   = .install_info
INSTALL_SUFFIX = Synch
UNINSTALL_DIR  = $(shell cat $(INSTALL_INFO))

# define any directories containing header files other than /usr/include
INCLUDES   = -Ilibconcurrent/includes -Ilibconcurrent

# define the output directory
BINDIR     = build/bin
LIBDIR     = build/lib
DOCSDIR    = build/docs
OBJDIR     = build/objects

LIBNAME    = libconcurrent.a
LIBNAMESO  = libconcurrent.so

BUILD_LOG=./build.log

# define the Cpp source files
PRIMITIVES = $(wildcard libconcurrent/primitives/*.cpp)
CONCURRENT = $(wildcard libconcurrent/concurrent/*.cpp)
BENCHMARKS = $(wildcard benchmarks/*bench.cpp)
SRCS       = $(PRIMITIVES) $(CONCURRENT) 

# define the Cpp object files 
OBJECTS    = $(SRCS:.cpp=.o)

# define the .run executable files
EXECS      = $(BENCHMARKS:.cpp=.run)

COLOR_CC   = [  \e[32mCC\e[39m  ]
COLOR_CP   = [  \e[32mCP\e[39m  ]
COLOR_AR   = [  \e[36mAR\e[39m  ]
COLOR_RM   = [  \e[31mRM\e[39m  ]
COLOR_ERR  = [ \e[31mERROR\e[39m ]

.PHONY: all clean

all: $(EXECS) 

$(EXECS): %.run : %.cpp libbuild
	@printf '$(COLOR_CC)\t$<\n'
	@$(COMPILER) $(CFLAGS) $(ARGS) $(INCLUDES) $(D_ARGS) $< -o $(BINDIR)/$(notdir $@) $(LIBDIR)/$(LIBNAME) $(LDLIBS)

.cpp.o:
	@printf '$(COLOR_CC)\t$<\n'
	@$(COMPILER) $(CFLAGS) $(ARGS) $(INCLUDES) $(D_ARGS) -c $< -o $(OBJDIR)/$(notdir $@)

libbuild: $(OBJECTS)
	@printf '$(COLOR_CC)\t$(LIBDIR)/$(LIBNAMESO)\n'
	@$(COMPILER) -shared -o $(LIBDIR)/$(LIBNAMESO) $(addprefix $(OBJDIR)/,$(notdir $(OBJECTS)))
	@printf '$(COLOR_AR)\t$(LIBDIR)/$(LIBNAME)\n'
	@$(AR) rcs $(LIBDIR)/$(LIBNAME) $(addprefix $(OBJDIR)/,$(notdir $(OBJECTS)))
```

**Comment by nkallima** [comment]
@cxk1686 Thanks for the reply,

Unfortunately, I have not access to any A72 machine.
Any more progress or info about running the scripts with smaller values for the `RUNS_PER_THREAD`.

**Comment by nkallima** [comment]
@cxk1686  
I recently had the opportunity to test the framework on a Raspberry Pi.
Please take a look at version 3.3.0 of the code, available at [v3.3.0](https://github.com/nkallima/sim-universal-construction/tree/v3.3.0-dev).
Although this version is still under development, we have addressed several bugs specifically related to machines with a weak memory model, such as the Raspberry Pi.

**Comment by nkallima** [comment]
Release https://github.com/nkallima/sim-universal-construction/releases/tag/v3.3.0 seems to solve these problems at least for me.
